design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,163 @@
+
+Increasing the number of CPU sockets will not only add CPUs to the guest but also increase the CPU resources available to the virt-launcher pod, via a migration (more on that in the Implementation section).
+The way LibVirt handles CPU hotplug is by exposing the maximum number of CPUs to the guest and turning off the unused CPUs. In this document, when we refer to "adding" CPUs to the guest, we're really just turning them on. This is why we need to define a maximum number of sockets in advance.
+
+## Drawbacks / Limitations
+- Since CPU hot-(un)plug involves a live migration, VMs will have to be `LiveMigratable` to enable CPU hotplug
+- Each disabled vCPU, so `(maxSockets - sockets) * cores * thread`, consumes 8MiB of overhead memory
+- This feature will be incompatible with CPU requests/limits, at least initially, since those values need to dynamically scale according to the current number of enabled CPUs
+
+## API Examples
+
+### VM Spec API
+
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  labels:
+    kubevirt.io/vm: vm-cirros
+  name: vm-cirros
+spec:
+  running: false
+  liveUpdateFeatures:
+    cpu
+      maximumSockets: 4
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-cirros
+    spec:
+      domain:
+        cpu:
+          sockets: 2
+          cores: 4
+          threads: 2
+```
+In this example, the user will be able to double the guest CPU resources by increasing the value of `sockets`.
+
+### VMI Spec API
+
+The VM definition above would translate to the following VMI:
+
+
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  labels:
+    special: vm-cirros
+  name: vm-cirros
+spec:
+  domain:
+    cpu:
+      maxSockets: 4
+      sockets: 2
+      cores: 4
+      threads: 2
+```
+
+### Status API
+
+The VMI status must indicate whether CPU hotplug is taking place. As the hotplug process begins the current state of `spec.domain.cpu` needs to be propagated back to the VM/VMI `status.currentCPUTopology` to allow the user to have an indication of the hotplug processing.
+
+During the hotplug process, the current CPU topology state will be reflected in VM/VMI `Status.currentCPUTopology` while the desired CPU topology state will be in `spec.Domain.CPU`. When the hotplug process completes the `Status.currentCPUTopology` will be cleared.
+
+```
+	// CPU allow specified the detailed CPU topology inside the vmi.
+		CurrentCPUTopology *CPU `json:"currentCPUTopology,omitempty"`
+```
+
+Conditions on the VM and VMI should indicate the status of the pending CPU hotplug actions. When errors occur, the conditions should propagate the errors to the VM.Status.Conditions field so users can receive feedback into the status of their declared state.
+
+```
+	// Indicates that the VMI is in progress of Hot vCPU Plug/UnPlug
+	VirtualMachineInstanceVCPUChange VirtualMachineInstanceConditionType = "HotVCPUChange"
```

Comment:
I would suggest to be more specific, referring to plug or unplug.

Reply:
Why would it matter? the main point is that the number of vCPUs in changing. Whether this is plug or unplug one could understand by looking at the vmi spec (desired state) and the currentCPUTopology (current state)

Reply:
Fair enough